### PR TITLE
Allow AccessControl filter to work without User component

### PR DESCRIPTION
**Pull request by bizley**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | a bit
| Tests pass?   | yes
| Fixed issues  | #4793

This allows `AccessControl` filter to work without `user` component since only matching role mechanism depends on it.

- `yii\filters\AccessControl::user` can be set to `false` to explicitly switch this component support off for the filter,
- `yii\filters\AccessControl::denyAccess()` throws `yii\web\ForbiddenHttpException` if the user is already logged in **or in case of detached User component**,
- `yii\filters\AccessRule::matchRole()` throws `yii\base\InvalidConfigException` if User component is detached and `yii\filters\AccessRule::roles` are not empty.

With default configuration behavior is the same as before.

Also `yiiunit\framework\filters\AccessRuleTest` methods don't require mocking User component anymore.

### CODE REVIEW

File: tests/framework/filters/AccessRuleTest.php [hunk 1]
```diff
@@ -180,10 +180,28 @@ public function testMatchRole($actionid, $allow, $userid, $expected) {
...
+    public function testMatchRoleWithoutUser() {
+        $action = $this->mockAction();
+        $request = $this->mockRequest();
+
+        $rule = new AccessRule([
+            'allow' => true,
+            'roles' => ['@'],
+        ]);
+
+        $this->setExpectedException('yii\base\InvalidConfigException');
```
Comment (SilverFire) [inline tests/framework/filters/AccessRuleTest.php]: Should be replaced with `expectException()`

https://github.com/yiisoft/yii2/blob/master/tests/compatibility.php#L18

**Review by SilverFire — CHANGES_REQUESTED** [review]
Generally good

**Review by SilverFire — APPROVED** [review]

**Review by samdark — APPROVED** [review]

File: framework/filters/AccessRule.php [hunk 2]
```diff
@@ -141,12 +142,16 @@ protected function matchController($controller)
...
      * @return bool whether the rule applies to the role
+     * @throws InvalidConfigException if User component is detached
      */
     protected function matchRole($user)
     {
         if (empty($this->roles)) {
             return true;
         }
+        if ($user === false) {
+            throw new InvalidConfigException('The required component is not specified.');
```
Comment (cebe) [inline framework/filters/AccessRule.php]: this error message could be improved:

> "The user application component must be available to specify roles in AccessRule."

**Review by cebe — CHANGES_REQUESTED** [review]

**Review by cebe — APPROVED** [review]

**Review by SilverFire — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**